File: src/core/backtest/backtestEngine.js
/**
 * backtestEngine.js — 期货回测引擎 (S4: BacktestEngine "dll")。
 * 组合核算：现金、保证金占用、持仓市值、浮动盈亏、权益；逐日盯市；
 * 展期滚动（主力切换自动换月）；交易成本（双边手续费 + 滑点）；保证金约束。
 * 撮合：信号 t 日生成，t+delay 日收盘成交（默认 delay=1，无前视）。
 */

export const DEFAULT_BACKTEST_CONFIG = {
  initialCapital: 10_000_000,
  commissionRate: 0.0002,   // 单边手续费率（按名义额）
  slippageTicks: 1,         // 单边滑点（跳）
  executionDelay: 1,        // 成交延迟（交易日）：1 = 次日收盘成交
  maxLeverage: 1.5,         // 总保证金 / 权益 上限
  useAdjPrice: true,        // 盈亏按后复权主连续价（消除展期跳空）
};

export class BacktestEngine {
  /**
   * @param {DataAccess} ds
   * @param {{rebalanceDates: string[], targets: Object<string,Object<string,number>>}} strategy
   * @returns 回测结果（权益曲线、交易明细、展期记录、每日快照）
   */
  run(ds, strategy, config = {}) {
    const cfg = Object.assign({}, DEFAULT_BACKTEST_CONFIG, config);
    const dates = ds.dates;
    const T = dates.length;
    const metaOf = (code) => ds.getMeta(code);
    const seriesCache = {};
    const S = (code) => (seriesCache[code] = seriesCache[code] || ds.getSeries(code));

    // 每个品种最后一个有效主连续价的下标（用于退市强平）
    const lastValidIdx = {};
    for (const code of ds.codes) {
      const s = S(code);
      let li = null;
      for (let t = 0; t < T; t++) if (s.mainAdj[t] != null) li = t;
      lastValidIdx[code] = li;
    }

    // 展期事件索引：code -> Map(date -> roll)
    const rollByDate = {};
    for (const code of ds.codes) {
      const m = new Map();
      for (const r of S(code).rolls) m.set(r.date, r);
      rollByDate[code] = m;
    }

    let cash = cfg.initialCapital;
    const positions = {}; // code -> {lots, dir, entryAdj, entryRaw, contract}
    const trades = [];
    const rolls = [];
    const equityArr = new Array(T).fill(null);
    const snapshots = new Array(T).fill(null);

    let pending = null; // {execIdx, targets}

    const legCost = (meta, price, lots) =>
      cfg.commissionRate * price * meta.mult * lots + cfg.slippageTicks * meta.tickValue * lots;

    // 计算当前浮动盈亏 / 权益 / 保证金占用 / 可用资金
    const stats = (t) => {
      let floatingPnL = 0;
      let usedMargin = 0;
      let gross = 0;
      for (const code of Object.keys(positions)) {
        const pos = positions[code];
        const s = S(code);
        const meta = metaOf(code);
        const adjT = s.mainAdj[t];
        const rawT = s.mainRaw[t];
        if (adjT != null) floatingPnL += pos.dir * (adjT - pos.entryAdj) * meta.mult * pos.lots;
        if (rawT != null) { usedMargin += pos.lots * rawT * meta.mult * meta.margin; gross += pos.lots * rawT * meta.mult; }
      }
      const equity = cash + floatingPnL;
      return { floatingPnL, usedMargin, gross, equity, available: equity - usedMargin };
    };

    // 退市强平：数据结束后按最后有效价平仓
    const liquidateDelisted = (t, date) => {
      for (const code of Object.keys(positions)) {
        const s = S(code);
        if (s.mainAdj[t] != null) continue;
        const li = lastValidIdx[code];
        if (li == null) { delete positions[code]; continue; }
        const pos = positions[code];
        const meta = metaOf(code);
        const adjP = s.mainAdj[li];
        const rawP = s.mainRaw[li] != null ? s.mainRaw[li] : pos.entryRaw;
        const pnl = pos.dir * (adjP - pos.entryAdj) * meta.mult * pos.lots;
        const cost = legCost(meta, rawP, pos.lots);
        cash += pnl - cost;
        trades.push({ date, code, side: 'close', dir: pos.dir, lots: pos.lots, price: rawP, adjPrice: adjP, notional: rawP * meta.mult * pos.lots, cost, pnl, contract: pos.contract, reason: 'delist' });
        delete positions[code];
      }
    };

    // 展期滚动：主力切换自动换月（平旧开新），盈亏连续（后复权），仅计换月成本
    const processRolls = (t, date) => {
      for (const code of Object.keys(positions)) {
        const roll = rollByDate[code] && rollByDate[code].get(date);
        if (!roll) continue;
        const pos = positions[code];
        const meta = metaOf(code);
        const oldRaw = roll.fromClose != null ? roll.fromClose : pos.entryRaw;
        const newRaw = roll.toClose != null ? roll.toClose : pos.entryRaw;
        const cost = legCost(meta, oldRaw, pos.lots) + legCost(meta, newRaw, pos.lots);
        cash -= cost;
        pos.contract = roll.to;
        pos.entryRaw = newRaw;
        rolls.push({ date, code, from: roll.from, to: roll.to, lots: pos.lots, cost });
      }
    };

    // 交易到目标（含手续费/滑点/已实现盈亏/平均成本）
    const tradeTo = (date, t, code, targetLots, targetDir, targetContract) => {
      const meta = metaOf(code);
      const s = S(code);
      const rawT = s.mainRaw[t];
      const adjT = s.mainAdj[t];
      const cur = positions[code];
      const curLots = cur ? cur.lots : 0;
      const curDir = cur ? cur.dir : 0;

      if (targetLots === 0 && curLots === 0) return;
      if (targetLots === 0) {
        const pnl = curDir * (adjT - cur.entryAdj) * meta.mult * curLots;
        const cost = legCost(meta, rawT, curLots);
        cash += pnl - cost;
        trades.push({ date, code, side: 'close', dir: curDir, lots: curLots, price: rawT, adjPrice: adjT, notional: rawT * meta.mult * curLots, cost, pnl, contract: cur.contract, reason: 'rebalance' });
        delete positions[code];
        return;
      }
      if (curLots === 0) {
        const cost = legCost(meta, rawT, targetLots);
        cash -= cost;
        positions[code] = { lots: targetLots, dir: targetDir, entryAdj: adjT, entryRaw: rawT, contract: targetContract };
        trades.push({ date, code, side: 'open', dir: targetDir, lots: targetLots, price: rawT, adjPrice: adjT, notional: rawT * meta.mult * targetLots, cost, pnl: 0, contract: targetContract, reason: 'rebalance' });
        return;
      }
      if (curDir === targetDir) {
        if (targetLots > curLots) {
          const add = targetLots - curLots;
          const cost = legCost(meta, rawT, add);
          cash -= cost;
          positions[code] = {
            lots: targetLots, dir: targetDir,
            entryAdj: (cur.lots * cur.entryAdj + add * adjT) / targetLots,
            entryRaw: (cur.lots * cur.entryRaw + add * rawT) / targetLots,
            contract: targetContract,
          };
          trades.push({ date, code, side: 'add', dir: targetDir, lots: add, price: rawT, adjPrice: adjT, notional: rawT * meta.mult * add, cost, pnl: 0, contract: targetContract, reason: 'rebalance' });
        } else if (targetLots < curLots) {
          const close = curLots - targetLots;
          const pnl = curDir * (adjT - cur.entryAdj) * meta.mult * close;
          const cost = legCost(meta, rawT, close);
          cash += pnl - cost;
          positions[code] = { lots: targetLots, dir: targetDir, entryAdj: cur.entryAdj, entryRaw: cur.entryRaw, contract: targetContract };
          trades.push({ date, code, side: 'reduce', dir: curDir, lots: close, price: rawT, adjPrice: adjT, notional: rawT * meta.mult * close, cost, pnl, contract: cur.contract, reason: 'rebalance' });
        }
        return;
      }
      // 方向翻转
      const pnlClose = curDir * (adjT - cur.entryAdj) * meta.mult * curLots;
      const costClose = legCost(meta, rawT, curLots);
      const costOpen = legCost(meta, rawT, targetLots);
      cash += pnlClose - costClose - costOpen;
      positions[code] = { lots: targetLots, dir: targetDir, entryAdj: adjT, entryRaw: rawT, contract: targetContract };
      trades.push({ date, code, side: 'flip', dir: curDir + '->' + targetDir, lots: curLots, lotsNew: targetLots, price: rawT, adjPrice: adjT, notional: rawT * meta.mult * (curLots + targetLots), cost: costClose + costOpen, pnl: pnlClose, contract: targetContract, reason: 'rebalance' });
    };

    // 执行调仓
    const executeRebalance = (t, date, targets) => {
      const st = stats(t);
      if (st.equity <= 0) { // 爆仓：全部平仓
        for (const code of Object.keys(positions)) tradeTo(date, t, code, 0, 0, null);
        return;
      }
      // 目标手数
      const desired = {};
      for (const code of Object.keys(targets)) {
        const w = targets[code];
        const meta = metaOf(code);
        const s = S(code);
        const rawT = s.mainRaw[t];
        if (rawT == null || !(rawT > 0)) continue;
        const notional = Math.abs(w) * st.equity;
        const lots = Math.floor(notional / (rawT * meta.mult));
        if (lots <= 0) continue;
        desired[code] = { lots, dir: w > 0 ? 1 : -1, contract: s.mainCode[date] };
      }
      // 保证金约束：总保证金 <= 权益 * maxLeverage
      let totalMargin = 0;
      for (const code of Object.keys(desired)) {
        const d = desired[code];
        const meta = metaOf(code);
        totalMargin += d.lots * S(code).mainRaw[t] * meta.mult * meta.margin;
      }
      const cap = st.equity * (cfg.maxLeverage || 1.5);
      if (totalMargin > cap) {
        const f = cap / totalMargin;
        for (const code of Object.keys(desired)) desired[code].lots = Math.floor(desired[code].lots * f);
      }
      // 交易到目标
      const allCodes = new Set([...Object.keys(positions), ...Object.keys(desired)]);
      for (const code of allCodes) {
        const d = desired[code];
        tradeTo(date, t, code, d ? d.lots : 0, d ? d.dir : 0, d ? d.contract : null);
      }
    };

    const rebSet = new Set(strategy.rebalanceDates);
    for (let t = 0; t < T; t++) {
      const date = dates[t];

      // 1) 退市强平
      liquidateDelisted(t, date);
      // 2) 展期滚动
      processRolls(t, date);
      // 3) 执行待成交调仓（收盘价成交）
      if (pending && pending.execIdx === t) {
        executeRebalance(t, date, pending.targets);
        pending = null;
      }
      // 4) 计划新调仓（信号 t 日生成，t+delay 成交）
      if (rebSet.has(date) && !pending) {
        const execIdx = Math.min(t + (cfg.executionDelay || 1), T - 1);
        pending = { execIdx, targets: strategy.targets[date] || {} };
      }

      // 5) 快照
      const st = stats(t);
      equityArr[t] = st.equity;
      snapshots[t] = {
        date,
        equity: st.equity,
        cash,
        floatingPnL: st.floatingPnL,
        usedMargin: st.usedMargin,
        available: st.available,
        grossNotional: st.gross,
        nav: st.equity / cfg.initialCapital,
        nPositions: Object.keys(positions).length,
      };
    }

    const final = stats(T - 1);
    const totalCost = trades.reduce((a, x) => a + (x.cost || 0), 0);
    const totalRollCost = rolls.reduce((a, x) => a + (x.cost || 0), 0);
    return {
      dates,
      equity: equityArr,
      snapshots,
      trades,
      rolls,
      positions,
      config: cfg,
      summary: {
        initialCapital: cfg.initialCapital,
        finalEquity: final.equity,
        totalCost,
        totalRollCost,
        nTrades: trades.length,
        nRolls: rolls.length,
      },
    };
  }
}

File: cs/StrategyCore/BacktestEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MockTrader.Data;

namespace MockTrader.Core
{
    public sealed class BacktestConfig
    {
        public double InitialCapital { get; set; } = 10_000_000;
        public double CommissionRate { get; set; } = 0.0002;
        public double SlippageTicks { get; set; } = 1;
        public int ExecutionDelay { get; set; } = 1;
        public double MaxLeverage { get; set; } = 1.5;
        public double ImpactCoef { get; set; } = 0;
        public int AdvWindow { get; set; } = 20;
        public double DrawdownCutoff { get; set; } = 0;
    }

    public sealed class Trade
    {
        public string Date { get; set; }
        public string Code { get; set; }
        public string Side { get; set; }
        public string Contract { get; set; }
        public string Reason { get; set; }
        public int Dir { get; set; }
        public int Lots { get; set; }
        public int LotsNew { get; set; }
        public string DirLabel { get; set; }
        public double Price { get; set; }
        public double AdjPrice { get; set; }
        public double Notional { get; set; }
        public double Cost { get; set; }
        public double Pnl { get; set; }
    }

    public sealed class RollRecord
    {
        public string Date { get; set; }
        public string Code { get; set; }
        public string From { get; set; }
        public string To { get; set; }
        public int Lots { get; set; }
        public double Cost { get; set; }
    }

    public sealed class Position
    {
        public int Lots { get; set; }
        public int Dir { get; set; }
        public double EntryAdj { get; set; }
        public double EntryRaw { get; set; }
        public string Contract { get; set; }
    }

    public sealed class Snapshot
    {
        public string Date { get; set; }
        public double Equity { get; set; }
        public double Cash { get; set; }
        public double FloatingPnL { get; set; }
        public double UsedMargin { get; set; }
        public double Available { get; set; }
        public double GrossNotional { get; set; }
        public double Nav { get; set; }
        public int NPositions { get; set; }
    }

    public sealed class BacktestSummary
    {
        public double InitialCapital { get; set; }
        public double FinalEquity { get; set; }
        public double TotalCost { get; set; }
        public double TotalRollCost { get; set; }
        public int NTrades { get; set; }
        public int NRolls { get; set; }
        public bool CircuitBroken { get; set; }
    }

    public sealed class BacktestResult
    {
        public List<string> Dates { get; set; }
        public double[] Equity { get; set; }
        public List<Snapshot> Snapshots { get; set; }
        public List<Trade> Trades { get; set; }
        public List<RollRecord> Rolls { get; set; }
        public BacktestSummary Summary { get; set; }
    }

    public sealed class BacktestEngine
    {
        public BacktestResult Run(DataAccess ds, StrategyResult strategy, BacktestConfig cfg = null)
        {
            cfg = cfg ?? new BacktestConfig();
            var dates = ds.Dates;
            int T = dates.Count;

            var seriesCache = new Dictionary<string, ContinuousSeries>();
            ContinuousSeries S(string code)
            {
                if (!seriesCache.TryGetValue(code, out var s)) { s = ds.GetSeries(code); seriesCache[code] = s; }
                return s;
            }

            var lastValidIdx = new Dictionary<string, int>();
            foreach (var code in ds.Codes)
            {
                var s = S(code);
                int li = -1;
                for (int t = 0; t < T; t++) if (s.MainAdj[t] != null) li = t;
                lastValidIdx[code] = li;
            }

            var rollByDate = new Dictionary<string, Dictionary<string, RollEvent>>();
            foreach (var code in ds.Codes)
            {
                var m = new Dictionary<string, RollEvent>();
                foreach (var r in S(code).Rolls) m[r.Date] = r;
                rollByDate[code] = m;
            }

            double cash = cfg.InitialCapital;
            var positions = new Dictionary<string, Position>();
            var trades = new List<Trade>();
            var rolls = new List<RollRecord>();
            var equityArr = new double[T];
            var snapshots = new Snapshot[T];

            (int execIdx, Dictionary<string, double> targets)? pending = null;

            double AdvLots(string code, int t)
            {
                var s = S(code);
                int w = cfg.AdvWindow > 0 ? cfg.AdvWindow : 20;
                double sum = 0;
                int cnt = 0;
                for (int i = Math.Max(0, t - w); i < t; i++)
                {
                    if (s.MainVol[i] != null) { sum += s.MainVol[i].Value; cnt++; }
                }
                return cnt > 0 ? sum / cnt : 0;
            }

            double LegCost(string code, int t, Variety meta, double price, int lots)
            {
                double c = cfg.CommissionRate * price * meta.Mult * lots + cfg.SlippageTicks * meta.TickValue * lots;
                if (cfg.ImpactCoef > 0)
                {
                    double adv = AdvLots(code, t);
                    if (adv > 0)
                    {
                        c += cfg.ImpactCoef * price * meta.Mult * lots * Math.Sqrt(lots / adv);
                    }
                }
                return c;
            }

            (double floating, double margin, double gross, double equity, double avail) Stats(int t)
            {
                double floating = 0, margin = 0, gross = 0;
                foreach (var kv in positions)
                {
                    var pos = kv.Value;
                    var s = S(kv.Key);
                    var meta = ds.GetMeta(kv.Key);
                    double? adjT = s.MainAdj[t];
                    double? rawT = s.MainRaw[t];
                    if (adjT != null) floating += pos.Dir * (adjT.Value - pos.EntryAdj) * meta.Mult * pos.Lots;
                    if (rawT != null)
                    {
                        margin += pos.Lots * rawT.Value * meta.Mult * meta.Margin;
                        gross += pos.Lots * rawT.Value * meta.Mult;
                    }
                }
                double equity = cash + floating;
                return (floating, margin, gross, equity, equity - margin);
            }

            void LiquidateDelisted(int t, string date)
            {
                foreach (var code in positions.Keys.ToList())
                {
                    var s = S(code);
                    if (s.MainAdj[t] != null) continue;
                    int li = lastValidIdx[code];
                    if (li < 0) { positions.Remove(code); continue; }
                    var pos = positions[code];
                    var meta = ds.GetMeta(code);
                    double adjP = s.MainAdj[li].Value;
                    double rawP = s.MainRaw[li] != null ? s.MainRaw[li].Value : pos.EntryRaw;
                    double pnl = pos.Dir * (adjP - pos.EntryAdj) * meta.Mult * pos.Lots;
                    double cost = LegCost(code, t, meta, rawP, pos.Lots);
                    cash += pnl - cost;
                    trades.Add(new Trade { Date = date, Code = code, Side = "close", Dir = pos.Dir, Lots = pos.Lots, Price = rawP, AdjPrice = adjP, Notional = rawP * meta.Mult * pos.Lots, Cost = cost, Pnl = pnl, Contract = pos.Contract, Reason = "delist" });
                    positions.Remove(code);
                }
            }

            void ProcessRolls(int t, string date)
            {
                foreach (var code in positions.Keys.ToList())
                {
                    if (rollByDate.TryGetValue(code, out var m) && m.TryGetValue(date, out var roll))
                    {
                        var pos = positions[code];
                        var meta = ds.GetMeta(code);
                        double oldRaw = roll.FromClose ?? pos.EntryRaw;
                        double newRaw = roll.ToClose ?? pos.EntryRaw;
                        double cost = LegCost(code, t, meta, oldRaw, pos.Lots) + LegCost(code, t, meta, newRaw, pos.Lots);
                        cash -= cost;
                        pos.Contract = roll.To;
                        pos.EntryRaw = newRaw;
                        rolls.Add(new RollRecord { Date = date, Code = code, From = roll.From, To = roll.To, Lots = pos.Lots, Cost = cost });
                    }
                }
            }

            void TradeTo(string date, int t, string code, int targetLots, int targetDir, string targetContract)
            {
                var meta = ds.GetMeta(code);
                var s = S(code);
                double rawT = s.MainRaw[t].Value;
                double adjT = s.MainAdj[t].Value;
                positions.TryGetValue(code, out var cur);
                int curLots = cur != null ? cur.Lots : 0;
                int curDir = cur != null ? cur.Dir : 0;

                if (targetLots == 0 && curLots == 0) return;
                if (targetLots == 0)
                {
                    double pnl = curDir * (adjT - cur.EntryAdj) * meta.Mult * curLots;
                    double cost = LegCost(code, t, meta, rawT, curLots);
                    cash += pnl - cost;
                    trades.Add(new Trade { Date = date, Code = code, Side = "close", Dir = curDir, Lots = curLots, Price = rawT, AdjPrice = adjT, Notional = rawT * meta.Mult * curLots, Cost = cost, Pnl = pnl, Contract = cur.Contract, Reason = "rebalance" });
                    positions.Remove(code);
                    return;
                }
                if (curLots == 0)
                {
                    double cost = LegCost(code, t, meta, rawT, targetLots);
                    cash -= cost;
                    positions[code] = new Position { Lots = targetLots, Dir = targetDir, EntryAdj = adjT, EntryRaw = rawT, Contract = targetContract };
                    trades.Add(new Trade { Date = date, Code = code, Side = "open", Dir = targetDir, Lots = targetLots, Price = rawT, AdjPrice = adjT, Notional = rawT * meta.Mult * targetLots, Cost = cost, Pnl = 0, Contract = targetContract, Reason = "rebalance" });
                    return;
                }
                if (curDir == targetDir)
                {
                    if (targetLots > curLots)
                    {
                        int add = targetLots - curLots;
                        double cost = LegCost(code, t, meta, rawT, add);
                        cash -= cost;
                        positions[code] = new Position { Lots = targetLots, Dir = targetDir, EntryAdj = (cur.Lots * cur.EntryAdj + add * adjT) / targetLots, EntryRaw = (cur.Lots * cur.EntryRaw + add * rawT) / targetLots, Contract = targetContract };
                        trades.Add(new Trade { Date = date, Code = code, Side = "add", Dir = targetDir, Lots = add, Price = rawT, AdjPrice = adjT, Notional = rawT * meta.Mult * add, Cost = cost, Pnl = 0, Contract = targetContract, Reason = "rebalance" });
                    }
                    else if (targetLots < curLots)
                    {
                        int close = curLots - targetLots;
                        double pnl = curDir * (adjT - cur.EntryAdj) * meta.Mult * close;
                        double cost = LegCost(code, t, meta, rawT, close);
                        cash += pnl - cost;
                        positions[code] = new Position { Lots = targetLots, Dir = targetDir, EntryAdj = cur.EntryAdj, EntryRaw = cur.EntryRaw, Contract = targetContract };
                        trades.Add(new Trade { Date = date, Code = code, Side = "reduce", Dir = curDir, Lots = close, Price = rawT, AdjPrice = adjT, Notional = rawT * meta.Mult * close, Cost = cost, Pnl = pnl, Contract = cur.Contract, Reason = "rebalance" });
                    }
                    return;
                }
                double pnlClose = curDir * (adjT - cur.EntryAdj) * meta.Mult * curLots;
                double costClose = LegCost(code, t, meta, rawT, curLots);
                double costOpen = LegCost(code, t, meta, rawT, targetLots);
                cash += pnlClose - costClose - costOpen;
                positions[code] = new Position { Lots = targetLots, Dir = targetDir, EntryAdj = adjT, EntryRaw = rawT, Contract = targetContract };
                trades.Add(new Trade { Date = date, Code = code, Side = "flip", Dir = curDir, Lots = curLots, LotsNew = targetLots, DirLabel = curDir + "->" + targetDir, Price = rawT, AdjPrice = adjT, Notional = rawT * meta.Mult * (curLots + targetLots), Cost = costClose + costOpen, Pnl = pnlClose, Contract = targetContract, Reason = "rebalance" });
            }

            void ExecuteRebalance(int t, string date, Dictionary<string, double> targets)
            {
                var st = Stats(t);
                if (st.equity <= 0)
                {
                    foreach (var code in positions.Keys.ToList()) TradeTo(date, t, code, 0, 0, null);
                    return;
                }
                var desired = new Dictionary<string, (int lots, int dir, string contract)>();
                foreach (var kv in targets)
                {
                    string code = kv.Key;
                    double w = kv.Value;
                    var meta = ds.GetMeta(code);
                    var s = S(code);
                    double rawT = s.MainRaw[t] ?? 0;
                    if (!(rawT > 0)) continue;
                    double notional = Math.Abs(w) * st.equity;
                    int lots = (int)Math.Floor(notional / (rawT * meta.Mult));
                    if (lots <= 0) continue;
                    desired[code] = (lots, w > 0 ? 1 : -1, s.MainCode[date]);
                }
                double totalMargin = 0;
                foreach (var kv in desired)
                {
                    var meta = ds.GetMeta(kv.Key);
                    totalMargin += kv.Value.lots * (S(kv.Key).MainRaw[t] ?? 0) * meta.Mult * meta.Margin;
                }
                double cap = st.equity * (cfg.MaxLeverage > 0 ? cfg.MaxLeverage : 1.5);
                if (totalMargin > cap)
                {
                    double f = cap / totalMargin;
                    foreach (var k in desired.Keys.ToList())
                    {
                        var d = desired[k];
                        desired[k] = ((int)Math.Floor(d.lots * f), d.dir, d.contract);
                    }
                }
                var allCodes = new HashSet<string>(positions.Keys);
                foreach (var k in desired.Keys) allCodes.Add(k);
                foreach (var code in allCodes)
                {
                    bool has = desired.TryGetValue(code, out var d);
                    TradeTo(date, t, code, has ? d.lots : 0, has ? d.dir : 0, has ? d.contract : null);
                }
            }

            var rebSet = new HashSet<string>(strategy.RebalanceDates);
            double peakEquity = cfg.InitialCapital;
            bool circuitBroken = false;
            for (int t = 0; t < T; t++)
            {
                string date = dates[t];
                LiquidateDelisted(t, date);
                ProcessRolls(t, date);
                if (pending != null && pending.Value.execIdx == t)
                {
                    ExecuteRebalance(t, date, pending.Value.targets);
                    pending = null;
                }
                if (rebSet.Contains(date) && pending == null && !circuitBroken)
                {
                    int execIdx = Math.Min(t + (cfg.ExecutionDelay > 0 ? cfg.ExecutionDelay : 1), T - 1);
                    var tg = strategy.Targets.TryGetValue(date, out var tmap) ? tmap : new Dictionary<string, double>();
                    pending = (execIdx, tg);
                }
                var st2 = Stats(t);
                equityArr[t] = st2.equity;
                snapshots[t] = new Snapshot
                {
                    Date = date, Equity = st2.equity, Cash = cash, FloatingPnL = st2.floating,
                    UsedMargin = st2.margin, Available = st2.avail, GrossNotional = st2.gross,
                    Nav = st2.equity / cfg.InitialCapital, NPositions = positions.Count
                };

                if (st2.equity > peakEquity) { peakEquity = st2.equity; }
                if (!circuitBroken && cfg.DrawdownCutoff > 0)
                {
                    double dd = peakEquity > 0 ? (peakEquity - st2.equity) / peakEquity : 0;
                    if (dd >= cfg.DrawdownCutoff)
                    {
                        foreach (var code in positions.Keys.ToList()) { TradeTo(date, t, code, 0, 0, null); }
                        circuitBroken = true;
                    }
                }
            }

            double totalCost = trades.Sum(x => x.Cost);
            double totalRollCost = rolls.Sum(x => x.Cost);
            return new BacktestResult
            {
                Dates = dates, Equity = equityArr, Snapshots = snapshots.ToList(), Trades = trades, Rolls = rolls,
                Summary = new BacktestSummary
                {
                    InitialCapital = cfg.InitialCapital, FinalEquity = equityArr[T - 1],
                    TotalCost = totalCost, TotalRollCost = totalRollCost, NTrades = trades.Count, NRolls = rolls.Count, CircuitBroken = circuitBroken
                }
            };
        }
    }
}
